File: Day5/day5.py
from copy import deepcopy
import sys

class Port:
    def __init__(self, cranes: str, mode_9001: bool = False):
        split_cranes = cranes.split("\n")
        stack_values = [element for element in split_cranes[-1] if element != " "]
        stacks = [[] for _ in stack_values]
        self.positions = [split_cranes[-1].index(value) for value in stack_values]

        for line in split_cranes[-2::-1]:
            for index in range(len(stacks)):
                contents = line[self.positions[index]]
                if contents != " ":
                    stacks[index].append(line[self.positions[index]])

        self.status = stacks
        self.mode_9001 = mode_9001

    def move(self, number: int, move_from: int, move_to: int):
        elements_to_move = self.status[move_from][-number:]
        del self.status[move_from][-number:]
        if not self.mode_9001:
            elements_to_move.reverse()
        self.status[move_to] += elements_to_move

    def print_top(self) -> str:
        top_elements = ""
        for stack in self.status:
            top_elements += stack[-1]
        return top_elements

    def represent_stack(self) -> str:
        stack_representation = ""
        for index in range(len(self.status)):
            stack_representation += f" {index + 1}  "
        finished = [False] * len(self.status)
        height = 0
        while not all(finished):
            current_line = ""
            for index in range(len(finished)):
                if len(self.status[index]) > height:
                    current_line += f"[{self.status[index][height]}] "
                else:
                    current_line += "    "
                    finished[index] = True
            stack_representation = current_line + "\n" + stack_representation
            height += 1
        return stack_representation


def extract_moves(moves: str) -> list[list[int]]:
    split_moves = moves.split("\n")
    split_moves.remove("")
    split_moves = [line.split(" ") for line in split_moves]
    split_moves = [
        [int(line[1]), int(line[3]) - 1, int(line[5]) - 1] for line in split_moves
    ]
    return split_moves


if __name__ == "__main__":
    try:
        file_name = sys.argv[1]
    except IndexError:
        file_name = "input.txt"
    cranes, moves = open(file_name).read().split(sep="\n\n")
    moves = extract_moves(moves)
    stacks_9000 = Port(cranes)
    stacks_9001 = deepcopy(stacks_9000)
    stacks_9001.mode_9001 = True
    for move in moves:
        stacks_9000.move(*move)
        stacks_9001.move(*move)

    print(
        f"The elements at the top after moving with CrateMover 9000 are {stacks_9000.print_top()}."
    )
    print(
        f"The elements at the top after moving with CrateMover 9001 are {stacks_9001.print_top()}."
    )

    #print("\nThis is how things are once CrateMover 9000 is done.")
    #print(stacks_9000.represent_stack())

    #print("\nThis is how things are once CrateMover 9001 is done.")
    #print(stacks_9001.represent_stack())

File: Day5/day5.rs
fn get_data() -> String {
    let input_args: Vec<String> = std::env::args().collect();
    let mut file_name: String = "input.txt".to_string();
    if input_args.len() > 1 {
        file_name = input_args[1].clone();
    }
    let mut data = std::fs::read_to_string(file_name).unwrap();
    if data.ends_with("\n") {
        data.pop();
    }
    return data;
}

fn extract_moves_from_instructions(instructions: &str) -> Vec<(usize, usize, usize)> {
    let mut count: usize = 0;
    let mut source: usize = 0;
    let mut destination: usize = 0;

    let mut moves = vec![];

    for instruction in instructions.lines() {
        for (idx, part) in instruction.split(' ').enumerate() {
            match idx {
                0 | 2 | 4 => continue,
                1 => count = part.parse().unwrap(),
                3 => source = part.parse().unwrap(),
                5 => destination = part.parse().unwrap(),
                _ => panic!("unexpected input: {:?}", instruction),
            }
        }
        moves.push((count, source - 1, destination - 1))
    }
    return moves;
}

fn extract_stacks(stacks_input: &str) -> Vec<Vec<char>> {
    let mut stacks = vec![vec![]];

    for line in stacks_input.lines() {
        if !line.starts_with(" 1") {
            for (idx, value) in line.as_bytes().chunks(4).enumerate() {
                if stacks.len() <= idx {
                    stacks.push(vec![]);
                }

                if value[1 as usize] != b' ' as u8 {
                    stacks[idx].push(value[1 as usize] as char);
                }
            }
        }
    }
    // We have the stacks but they are reversed
    for stack in stacks.iter_mut() {
        stack.reverse();
    }
    return stacks;
}

fn perform_moves(
    instructions: &[(usize, usize, usize)],
    mut stacks: Vec<Vec<char>>,
    mode_9001: bool,
) -> Vec<Vec<char>> {
    for (count, source, destination) in instructions.iter().copied() {
        let split_point = stacks[source].len() - count;
        let mut crates_to_move = stacks[source].split_off(split_point);

        if mode_9001 == false {
            crates_to_move.reverse();
        }

        stacks[destination].append(&mut crates_to_move);
    }
    return stacks;
}

fn get_top_elements(stacks: &Vec<Vec<char>>) -> String {
    let mut answer = String::with_capacity(stacks.len());
    for stack in stacks {
        answer.push(stack.last().unwrap().clone());
    }
    return answer;
}

fn main() {
    let data = get_data();
    let stacks_and_instructions: Vec<&str> = data.split("\n\n").collect();
    let mut stacks_9000 = extract_stacks(stacks_and_instructions[0]);
    let mut stacks_9001 = stacks_9000.clone();
    let instructions = extract_moves_from_instructions(stacks_and_instructions[1]);

    stacks_9000 = perform_moves(&instructions, stacks_9000, false);
    stacks_9001 = perform_moves(&instructions, stacks_9001, true);

    let answer_part_1 = get_top_elements(&stacks_9000);
    let answer_part_2 = get_top_elements(&stacks_9001);

    println!(
        "The elements at the top after moving with CrateMover 9000 are {}",
        answer_part_1
    );
    println!(
        "The elements at the top after moving with CrateMover 9001 are {}",
        answer_part_2
    );
}
